src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,561 @@
+            Json::Object(ref d) => self.encode_obj(d),
+            Json::Array(ref d) => self.encode_array(d),
+            Json::Literal(d) => self.encode_literal(d),
+            Json::I64(d) => self.encode_json_i64(d),
+            Json::Double(d) => self.encode_json_f64(d),
+            Json::String(ref d) => self.encode_str(d),
+        }
+    }
+
+    fn encode_obj(&mut self, data: &BTreeMap<String, Json>) -> Result<()> {
+        // object: element-count size key-entry* value-entry* key* value*
+        // element-count ::= uint32 number of members in object or number of elements in array
+        let element_count = data.len();
+        let element_count_len = LENGTH_U32;
+        // size ::= uint32 number of bytes in the binary representation of the object or array
+        let size_len = LENGTH_U32;
+        // key-entry ::= key-offset(uint32) key-length(uint16)
+        let key_entries_len = LENGTH_KEY_ENTRY * element_count;
+        // value-entry ::= type(byte) offset-or-inlined-value(uint32)
+        let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+        let mut key_entries = Vec::with_capacity(key_entries_len);
+        let mut encode_keys = Vec::new();
+        let mut key_offset = element_count_len + size_len + key_entries_len + value_entries_len;
+        for key in data.keys() {
+            let encode_key = key.as_bytes();
+            let key_len = try!(encode_keys.write(encode_key));
+            try!(key_entries.encode_u32_le(key_offset as u32));
+            try!(key_entries.encode_u16_le(key_len as u16));
+            key_offset += key_len;
+        }
+
+        let mut value_offset = key_offset as u32;
+        let mut value_entries = Vec::with_capacity(value_entries_len);
+        let mut encode_values = vec![];
+        for value in data.values() {
+            try!(value_entries.encode_json_item(value, &mut value_offset, &mut encode_values));
+        }
+        let size = value_offset;
+        try!(self.encode_u32_le(element_count as u32));
+        try!(self.encode_u32_le(size as u32));
+        try!(self.write_all(key_entries.as_mut()));
+        try!(self.write_all(value_entries.as_mut()));
+        try!(self.write_all(encode_keys.as_mut()));
+        try!(self.write_all(encode_values.as_mut()));
+        Ok(())
+    }
+
+    fn encode_array(&mut self, data: &[Json]) -> Result<()> {
+        // array ::= element-count size value-entry* value*
+        let element_count = data.len();
+        let count_len = LENGTH_U32;
+        let size_len = LENGTH_U32;
+        let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+        let mut value_offset = (count_len + size_len + value_entries_len) as u32;
+        let mut value_entries = Vec::with_capacity(value_entries_len);
+        let mut encode_values = vec![];
+        for value in data {
+            try!(value_entries.encode_json_item(value, &mut value_offset, &mut encode_values));
+        }
+        let total_size = value_offset;
```

Comment:
I think `total_size` should be `encode_values.len() + count_len + size_len + value_entries_len`.